ctdcal/fit_ctd.py

```diff
@@ -15,6 +15,9 @@
 import json
 from pathlib import Path
 import config as cfg
+import matplotlib.pyplot as plt
```

Comment:
I would take a second think about whether or not pyplot should be imported and called here, rather than put into another module. I think you can argue either way for it. The original structure was set to be:

ctdcal/ctdcal: libraries, API, anything that wouldn't be called on its own
ctdcal/scripts: files that could be run on their own to use ctdcal

Whether it makes more sense to have the plots integrated here or separated off because other areas start using them may be something that will only be proven in time.

Reply:
I suppose ctdcal/ctd_plots.py might be a better location for residual_plot() to live. I'm also imagining giving that a thorough cleanup and implementing things I learned from this round of new figs. Let's chat about this on zoom next time we call.

I'll leave this un-merged for now and circle back

Reply:
Another item for the big ol todo pile (when you do the cleanup)... also keep in mind that I haven't been part of the IRL conversations

If possible, I think you'd want to architecture this so that matplotlib (or any other plotting library) is an optional dependency. The idea being that the data processing/calibration parts should be able to work without the needing what tends to be the "more complicated" to install deps. Though there might be an iterative workflow path that requires looking at a plot, adjusting some param, rerun cal, looking at a plot, loop...

As an example for a "headless" mode, when the old ODF software was set up by someone who really knew what they were doing, it would process the CTD cast immediately after completion using the previous cal configs so data was always available with the "best current" calibration even when the CTD processor was sleeping. Though you could argue that making some plots à la UHDAS could be part of this automatic workflow...

Reply:
One way I could see this working would be to replace all of the plot steps with exporting data to .csv (or .nc, .pkl, whatever) and then an optional call to the plotting module to visualize the data. My other thought is that plotting (at this stage of processing) could likely be fully replaced with some statistical checks to determine whether or not the fit parameters are good/questionable/bad.

Regardless of how specifically this is implemented, it's a good thing to keep in mind so thanks for bringing it up.

Reply:
To build off of @DocOtak talking about [optional dependencies](https://www.python.org/dev/peps/pep-0508/#extras) we could simply generate files that could be loaded nicely into ODV, or some other plotting software, without requiring something like matplotlib be installed. I think this would be easiest for the future with xarray, but we have the machinery for WHP-exchange right now, although it'll need to be hand tuned for every cruise.

But getting back on point, the plotting packages (matplotlib, bokeh) are seen as a critical part of ctdcal as a whole. You can't escape [Anscombe's quartet](https://en.wikipedia.org/wiki/Anscombe%27s_quartet) and it's implications. Whether or not we want to put effort into making a low bandwidth-friendly installation by minimizing dependencies (pandas is still heavy iirc), how to make that easy for others to install, is something to be worked out.

Actually getting on point, I think keeping the plotting routines off to the side, and putting them in an exception in case mpl is probably a good idea. Keep the core backend code separate from the visualization. Having a separate module for statistical tests on the data is something sorely needed, but something that can wait. Figuring out how to build files that can be nicely dumped out of xarray instead of having to hand write headers for WHP-exchange gets to the core of your thought, and fixes a lot of other problems we have too.